qualtran/bloqs/chemistry/thc.ipynb

```diff
@@ -0,0 +1,291 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "8a1cab3e",
+   "metadata": {
+    "cq.autogen": "title_cell"
+   },
+   "source": [
+    "# Tensor Hypercontraction\n",
+    "\n",
+    "SELECT and PREPARE for the molecular tensor hypercontraction (THC) hamiltonian"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "76fe66a3",
+   "metadata": {
+    "cq.autogen": "top_imports"
+   },
+   "outputs": [],
+   "source": [
+    "from qualtran import Bloq, CompositeBloq, BloqBuilder, Signature, Register\n",
+    "from qualtran.drawing import show_bloq\n",
+    "from typing import *\n",
+    "import numpy as np"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1423d573",
+   "metadata": {
+    "cq.autogen": "_make_uniform_superposition.md"
+   },
+   "source": [
+    "## `UniformSuperpositionTHC`\n",
+    "Prepare uniform superposition state for THC.\n",
+    "\n",
+    "$$\n",
+    "    |0\\rangle^{\\otimes 2\\log(M+1)} \\rightarrow \\sum_{\\mu\\le\\nu}^{M} |\\mu\\rangle|\\nu\\rangle + \\sum_{\\mu}^{N/2}|\\mu\\rangle|\\nu=M+1\\rangle,\n",
+    "$$\n",
+    "\n",
+    "where $M$ is the THC auxiliary dimension, and $N$ is the number of spin orbitals.\n",
+    "\n",
+    "The toffoli complexity of this gate should be 10 * log(M+1) + 2 b_r - 9.\n",
+    "Currently it is a good deal larger due to:\n",
+    "    1. inverting inequality tests should not need more toffolis.\n",
+    "    2. We are not using phase-gradient gate toffoli cost for Ry rotations\n",
+    "\n",
+    "#### Parameters\n",
+    " - `num_mu`: THC auxiliary index dimension $M$\n",
+    " - `num_spin_orb`: number of spin orbitals $N$ \n",
+    "\n",
+    "Registers:\n",
+    "- mu: $\\mu$ register.\n",
+    "- nu: $\\nu$ register.\n",
+    "- succ: ancilla flagging success of amplitude amplification.\n",
+    "- eq_nu_mp1: ancillas for flagging if $\\nu = M+1$.\n",
+    "\n",
+    "#### References\n",
+    "[Even more efficient quantum computations of chemistry through tensor hypercontraction](https://arxiv.org/pdf/2011.03494.pdf). Eq. 29.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1b6bb0e0",
+   "metadata": {
+    "cq.autogen": "_make_uniform_superposition.py"
+   },
+   "outputs": [],
+   "source": [
+    "from qualtran.bloqs.chemistry.thc import UniformSuperpositionTHC\n",
+    "\n",
+    "num_mu = 10\n",
+    "num_spin_orb = 4\n",
+    "uniform_bloq = UniformSuperpositionTHC(num_mu=num_mu, num_spin_orb=num_spin_orb)\n",
+    "show_bloq(uniform_bloq)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a89e358f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qualtran.resource_counting import get_bloq_counts_graph\n",
+    "from qualtran.drawing import GraphvizCounts\n",
+    "from qualtran.bloqs.chemistry.thc_notebook_utils import generalize\n",
+    "graph, sigma = get_bloq_counts_graph(uniform_bloq, generalizer=generalize)\n",
```

Comment:
this is now broken, possibly by the new cirq stuff

```
ValueError: Cannot decompose from cirq op: CirqGateAsBloq(gate=cirq.X) doesn't have a cirq op.
```

can you work with @tanujkhattar to come up with a fix?

Reply:
Here is a fix https://github.com/quantumlib/Qualtran/pull/419